src/bundler.ts

```diff
@@ -102,346 +97,268 @@ export class Bundler {
     this.excludes = Array.isArray(opts.excludes) ? opts.excludes : [];
     this.stripComments = Boolean(opts.stripComments);
     this.enableCssInlining = Boolean(opts.inlineCss);
     this.enableScriptInlining = Boolean(opts.inlineScripts);
-    this.inputUrl =
-        String(opts.inputUrl) === opts.inputUrl ? opts.inputUrl : '';
   }
 
-
   /**
-   * Return the URL this import should point to in the given bundle.
+   * Given urls to entrypoint html documents, produce a collection of bundled
+   * documents with HTML imports, external stylesheets and external scripts
+   * inlined according to the options for this Bundler.
    *
-   * If the URL is part of the bundle, this method returns `true`.
-   *
-   * If the URL is part of another bundle, this method returns the url of that
-   * bundle.
-   *
-   * If the URL isn't part of a bundle, this method returns `false`
+   * @param {Array<string>} entrypoints The list of entrypoints that will be
+   *     analyzed for dependencies. The results of the analysis will be passed
+   *     to the `strategy`.
+   * @param {BundleStrategy} strategy The strategy used to construct the
+   *     output bundles. See 'polymer-analyzer/src/bundle-manifest'.
+   * @param {BundleUrlMapper} mapper A function that produces urls for the
+   *     generated bundles. See 'polymer-analyzer/src/bundle-manifest'.
    */
-  resolveBundleUrl(
-      url: string,
-      bundle: AssignedBundle,
-      manifest: BundleManifest): boolean|string {
-    const targetBundle = manifest.getBundleForFile(url);
-    if (!targetBundle || !targetBundle.url) {
-      return false;
-    }
-    if (targetBundle.url !== bundle.url) {
-      const relative = urlUtils.relativeUrl(bundle.url, targetBundle.url);
-      if (!relative) {
-        throw new Error('Unable to compute relative url to bundle');
-      }
-      return relative;
+  async bundle(
+      entrypoints: string[],
+      strategy?: BundleStrategy,
+      mapper?: BundleUrlMapper): Promise<DocumentCollection> {
+    // TODO(usergenic): Expose the generateBundleManifest operation
+    // as public and make the interface for bundle method take a manifest.
+    const manifest =
+        await this._generateBundleManifest(entrypoints, strategy, mapper);
+
+    const bundledDocuments: DocumentCollection =
+        new Map<string, BundledDocument>();
+
+    for (const bundleEntry of manifest.bundles) {
+      const bundleUrl = bundleEntry[0];
+      const bundle = {url: bundleUrl, bundle: bundleEntry[1]};
+      const bundledAst =
+          await this._bundleDocument(bundle, manifest, bundle.bundle.files);
+      bundledDocuments.set(
+          bundleUrl, {ast: bundledAst, files: Array.from(bundle.bundle.files)});
     }
-    return true;
+
+    return bundledDocuments;
   }
 
-  isStripExcludedHref(url: string): boolean {
-    if (!this.stripExcludes) {
-      return false;
+  /**
+   * Add HTML Import elements for each file in the bundle.  We append all the
```

Comment:
Seems like we need to be careful to define what happens when we have a graph of imports in documents from an existing root of a bundle, and a new authoritative set of document that the bundle contains from the manifest. Appending imports would be one way of implementing   that. Another bundler would want to match the behavior even if they didn't do it by appending imports.

The important aspect of that is ordering. It seems like this approach preserves an important property - that the order of imports in actual files is respected first, and only if an import isn't already transitively reachable from an existing bundle root is it's ordering determined by the manifest.

I would slightly and eventually like to see this done without injecting imports, since imports that aren't part of the bundle are moved on the fly - seems like additional imports should be able to be added after inlining everything and seeing what's left. But that's a very low priority wish.